### DISCUSSION TITLE
Schema for point grids

### DISCUSSION THREAD

**Discussion by ptgrogan** [eose-tools-org/eose-api#2] [General]
TAT-C has a [hidden schema](https://github.com/code-lab-org/tatc/blob/main/app/tatc_app/generation/schemas.py) for a "point generator" used to create point grids.

Simplifying the code below, attributes are: 

- `method` (either fibonacci_lattice or cubed_sphere), 
- `distance` (typical distance between points; converted to angular measure for cubed sphere), 
- `elevation` to set the z-coordinate of all generated points, and 
- `mask` to define a spatial region over which to generate points. Note that mask allows either a GeoJSON Polygon, a hard-coded known shape loaded from a shapefile (only "conus" is available), and three-letter country abbreviations also loaded from a shapefile.

```
class PointGeneratorMethod(str, Enum):
    """
    Method for point generation.
    """

    fibonacci_lattice = "fibonacci_lattice"
    cubed_square = "cubed_square"


class KnownShape(str, Enum):
    """
    Known shapes to use as masks.
    """

    conus = "conus"


class PointGenerator(APIModel):
    """
    Specification to generate points.
    """

    method: PointGeneratorMethod = Field(
        PointGeneratorMethod.cubed_square, description="Point generation method."
    )
    distance: float = Field(
        1e6, description="Characteristic distance (meters) between points."
    )
    elevation: float = Field(
        0,
        description="Elevation (meters) above datum in the WGS 84 coordinate system.",
    )
    mask: Optional[
        Union[Polygon, KnownShape, constr(min_length=3, max_length=3)]
    ] = Field(
        None,
        description="Mask to limit the extent of generated points. Allows ISO 3166-1 alpha-3 country codes.",
        example=Polygon.parse_obj(mapping(box(-180, -90, 180, 90))),
    )
```

**Comment by jsipps26** [comment]
Are point grids necessarily spread on the surface of a sphere or are we going to have the option to spread by equal area over the ellipsoid, such as WGS84? This is slightly different than just specifying elevation above WGS84, since the lat/lon locations of points will change (slightly).

Something I realized during our last discussion is that points are usually equally spaced by area (or so I understand), which means that for any aggregation, output values are unbiased. However, I'm not sure if the points actually need to be equally spaced by area - they can just be weighted by area. So this would require knowing what area each point occupies, and require that the Point object intrinsically contains an "area" value. This however doesn't make much sense for a general Point object attributable to a spacecraft or ground station... so maybe area metrics could be present only in the PointGrid class.

**Reply by ptgrogan** [reply]
The typical grid spacing used within the Earth science community is equal spacing in latitude/longitude (cubed sphere) which is not equally spaced in area. I am not aware of any technique that actually spaces equal in area, although the Fibonacci lattice performs better than cubed sphere.

You are right that we can use weighting functions to retrieve spatial metrics; however, area would be a property of a cell (or other polygon), not a point. For cubed sphere, one can imagine each point as describing a corner or center of a grid cell. A point array should be compatible with this perspective but not necessarily require it.

**Reply by vinay0000** [reply]
I think Jonathan is referring to EASE grids. Remote sensing (observational) science data is commonly available defined on this grid, and we have come across this in the D-SHIELD project also. During coverage calculations we calculated coverage over this grid.

**Reply by jsipps26** [reply]
I think the simplest is to assume the coverage grid will be equally spaced by area, and that area is derivable knowing how many points there are and the area over which coverage is computed (so the "point-based" area quantity need not be saved in the point groups). Equal-spaced grids are possible by making equidistant points at centers of grid cells on an equal-area projection, such as sinusoidal equal-area (used by MODIS), cylindrical equal-area (EASE, used by GEDI), or lambert azimuthal equal-area. One can even distribute points across the ellipsoid, such that the area between points is approximately the same.

**Reply by ryanketzner** [reply]
There are equal-area grids out there (I'm not familiar with EASE but I have seen lots of others, Healpix is one example). But all of these that I am aware of have some precise definition for what constitutes a "cell", and it is these cells which have equal area on the sphere. I don't think "equal area" could be precisely defined without providing a cell boundary definition, though the general idea makes sense (roughly uniform point density on the sphere). Some people use "cells" like this as a fundamental coverage calculation unit (I spoke to some people from Planet at IEEE aero who mentioned that they do this), but this seems like a more advanced feature which would be separate from the more basic and commonly used point grid.

I definitely agree that area should be a property of a cell (which is presumably defined as a polygon) and not of a point/point grid. For convenience, I see how it would be nice to have an "area" for pure point grids, which assumes each point in an n-point grid has an associated "area" of 4pi sr/n, but this could get thorny quickly, since it would also be common to specify grids which have very nonuniform density (i.e., a grid with uniform lat/lon spacing). So, it might be easier to specify these kinds of properties downstream if needed, alongside the grid, in the coverage metrics calculation software for example. Also, even if the _basic_ grid schemata do not have a concept of "area", it would still be possible to define a "uniform grid" concept later on if desired, which extends the basic grid to include this concept.

**Reply by ptgrogan** [reply]
Good to learn about some of the equal area approaches, thanks! 

I suspect the specification of point grids will have to be tied, in some way, to different methods. In the TAT-C example, the discriminator "method" does this and, coincidentally, both options take the same inputs. Arguably, cubed sphere may be better parameterized by a delta-longitude and delta-latitude rather than distance (currently, delta-longitude and delta-latitude are computed based on distance).

I agree with @ryanketzner  to consider attributes such as point area as a downstream item potentially associated with cells (polygons) rather than points (TAT-C has a similar distinction).

**Comment by vinay0000** [comment]
I think the another necessary (and probably more required) functionality to be supported, is to provide for a interface for the user to provide points from a file regardless of whether there is a 'structure' to it or not. (By 'structured points' what I mean are points 'generated' using an algorithm.)
These points would be specified in a specific coordinate system by the user.
The burden of generating the points (whether equal area spacing, or equal angular spacing, ...., and whether it is on a sphere or a ellipsoid) is upon the user. 

Regarding the coordinate systems, we can support the user to specify either the WGS84 geocentric (spherical) and the WGS84 geodetic (ellipsoidal). A conversion function between these two can be written.

We need to define the format of the list of points in the 'file'. Currently in EO-Sim I have a ``path`` key which points to a ``csv`` file, where the columns (labels, value type, units, range), header rows are defined to be in a specific format.  

@ryanketzner

**Reply by ptgrogan** [reply]
If we do a good job defining an object schema, it should be easy to serialize/deserialize from any other source including file format. It is generally difficult to rely on external data sources for an object schema (for example, how to keep track of relative file paths when submitting an HTTP request?), so I would recommend deserializing from file prior to instantiating an object. In other words, we want have all of the information in one document. I generally oppose defining a format for external files because it is simply not portable.

I'm not sure what you mean by WGS84 geocentric (spherical) versus WGS84 geodetic (ellipsoidal). My understanding is that WGS84 defines an Earth-centered, Earth-fixed frame (i.e., coordinate system axes) and an ellipsoid having a specific surface geometry. Perhaps the difference in the radial dimensions interpreted from surface versus the origin?

I want to emphasize that we are building an interchange format, and it is desirable to constrain it as much as possible because everything left as an option turns into a requirement for implementation. Yes, we may be able to provide a reference implementation in one language (e.g., Python), but, for example, if someone wants to build a JavaScript compliant application, they will have to implement all of the bells and whistles to be "compliant."

**Reply by vinay0000** [reply]
I agree that it would not be good to enforce a common file format and avoid specifying relative paths. The drawback is that the file could become data "heavy".
I think for mission data purposes this could be OKish. We have dealt with applications in which the point-grid files which weigh several Gigabytes (resolution in the order of few kilometers over global land-surface). If we de-serialize the data and read in the point-grid into python dictionaries, the dynamic memory allocation which would be required is huge. 

You suggested deserializing from file prior to instantiating an object. I would actually consider the opposite strategy. Given a point-grid specification in the ``eose-api`` format, we could serialize it (to whatever format our tools such as VASSAR/EO-Sim/ TAT-C support) and then read it in (say as a ``numpy`` arrray). Here the "serializer" also plays the role of "transformer", and required dynamic memory is small. Ofcourse, if the tool reads in objects directly, there is no need for the serializer.
Similarly when writing out the results, the serialized data can be written out onto the filesystem or a dynamic memory location, and then a "post-processor" could convert it to a de-serialized format.
(Am I missing something here?)

Related to this, what would be the strategy for specifying "nature" files, i.e. science data? 

> I'm not sure what you mean by WGS84 geocentric (spherical) versus WGS84 geodetic (ellipsoidal). 
WGS84 geocentric refers to [EPSG 4328](https://spatialreference.org/ref/epsg/4328/) and WGS84 geodetic refers to [EPSG 4326](https://epsg.io/4326). 

Writing "spherical" next to WGS84 geocentric was a mistake. We use spherical Earth assumptions at several places during coverage and data-metrics calculations. During these calculations we use [geocentric coordinates (latitude)](https://en.wikipedia.org/wiki/Latitude#Geocentric_latitude) and the mean (WGS84) radius of Earth.
I am keen to (somehow) support a spherical Earth model, since many of the mission calculations assume this.

**Reply by ryanketzner** [reply]
I also agree that a common file format shouldn't be enforced, but it might still be a good idea to have a "default" format for input files. The user is always free (and encouraged to) write their own file formats for input, but there is also a built-in format for ease of use. In this sense, the user would not be required to even use the default file format in order to be compliant, but they would have the option to do so. 

A question: is there even any way to *prohibit* the user from de-serializing directly to an object (by defining their own functions to do so)? To me it seems like regardless, the user would be able to write a "makeGrid" function which could be arbitrarily complex (for example, it could even be that the grid is specified in a custom binary format, and the user has written a custom utility to transform this directly into a numpy array, or whatever the underlying data storage is for the Grid class). I think this is always possible because Python doesn't have private data members. But I am not an expert in Python.

**Reply by ptgrogan** [reply]
My previous comment may have been misunderstood -- I have nothing against serializing data -- I would like to avoid having file paths be a part of the interface specification.

**Reply by ptgrogan** [reply]
I'll also reiterate the difference between the format required by an interface versus formats supported within a tool. For example, if the interface requires coordinates represented in EPSG:4326, it does not require all tools to use EPSG:4326 for all coordinates; but rather provide a transformer for EPSG:4326 to/from any internal formats used. It is essential to 1) ensure we have all the information _in the interface_ required to perform transformations of interest, and 2) specify all specific formats of particular interest to be included (required) _in the interface_.

**Comment by ptgrogan** [comment]
I posted a branch with an initial UniformAngularGrid object model: https://github.com/eose-tools-org/eose-api/tree/dev-points
